keps/2937-resource-transformer/README.md

```diff
@@ -0,0 +1,617 @@
+transformed resource list is just augmented with the computed `credits` resource.
+Most likely `credits` would be in a separate ResourceGroup than the "real" resources (similar to the
+[example of a ClusterQueue with software licenses](https://kueue.sigs.k8s.io/docs/concepts/cluster_queue/#resource-groups))
+and the `borrowingLimit` and `lendingLimit` for
+`credits` would be set to `0` to enforce the intended fiscal constraint.
+
+For example:
+* (`foo.com/gpu`, `example.com/credits`, `10`, `retain`)
+* (`cpu`, `example.com/credits`, `1`, `retain`)
+
+```yaml
+  - coveredResources: ["cpu", "memory", "foo.com/gpu"]
+    flavors:
+    - name: "spot"
+      resources:
+      - name: "cpu"
+        nominalQuota: 10
+      - name: "memory"
+        nominalQuota: 36Gi
+      - name: "foo.com/gpu"
+        nominalQuota: 40
+    - name: "on-demand"
+      resources:
+      - name: "cpu"
+        nominalQuota: 18
+      - name: "memory"
+        nominalQuota: 72Gi
+      - name: "foo.com/gpu"
+        nominalQuota: 100
+  - coveredResources: ["example.com/credits"]
+    flavors:
+    - name: "team1-budget"
+      resources:
+      - name: "example.com/credits"
+        nominalQuota: 750
+        borrowingLimit: 0
+        lendingLimit: 0
+```
+
+With the single global `ResourceMapping` proposed in this KEP, this use case may
+be of limited applicability.  It would benefit from finer-grained
+[alternatives](#alternatives) that would for example `cpu` allocated from
+the `spot` flavor to cost fewer credits than when allocated from the `on-demand` flavor.
+
+## Design Details
+
+Since Kueue's workload.Info struct is already a well-defined
+place where resource filtering is being applied to compute the effective resource
+request from a Workload's PodSpec, the core of the implementation is
+straightforward and follows from how `excludedResourcePrefixes` are already processed.
+
+The non-trivial design issues are how to configure the transformations,
+the possible granularity of that configuration, and how to maximize observability
+of the effective resources requested by a Workload.
+
+### Specifying the Transformation
+
+In this KEP we propose to support a global set of resource transformations that would
+be applied to all Workloads.  A `ResourceMapping` instance would be added to Kueue's
+`Configuration` API as part of the existing `Resources` struct.
+More [flexible alternatives](#alternatives) were considered, but left as possible
+future extensions.
+
+The definitions below would be added to Kueue's `Configuration` types
+```go
+type Resources struct {
+	// ExcludedResourcePrefixes defines which resources should be ignored by Kueue
+	ExcludeResourcePrefixes []string `json:"excludeResourcePrefixes,omitempty"`
+
+	// Transformations defines how to transform PodSpec resources into Workload resource requests.
+	// +listType=map
+	// +listMapKey=input
+	Transformations []ResourceMapping `json:"transformations,omitempty"`
```

Comment:
nit, but I would suggest to align the field name with the type name. Maybe call the type "ResourceTransformation" - I think it will play nicely. Rename ResourceMappingStrategy consistently in that case. Alternative is to call the field "ResourceMappings", I'm good with both.

Reply:
Makes sense.   Done.